Make jsviewer table have a sortable index

eteq opened the pull request:
I noticed something while playing around with the new `show_in_notebook` functionality: the jsviewer view into a table is *not* in the same order that the table is in in python.  This makes sense on reflection, because `DataTables` allows sorting, and seems to default to sorting on the first column.  But of course that's not always the "in-memory" order.

So this PR makes a simple modification to create a "fake" `row_index` column both with `show_in_notebook` and `show_inbrowser(jsviewer=True)`.  There are no tests or anything because I didn't want to spend too much time on it if others thing we don't want to do this.  It might also make more sense to actually implement this at a lower level, like where the HTML gets generated - this just adds in a column and then removes it immediately after the HTML is generated, but that feels a bit hacky.

@taldcroft @mhvk @saimn - what do you think?

(Screenshot below demonstrates how the notebook looks with and without this PR)
![prepost_virt_col](https://cloud.githubusercontent.com/assets/346587/11827323/36a4b78e-a359-11e5-8090-6396d4d1d74c.png)

taldcroft commented:
Don't we just want to have the initial sorting disabled?  I haven't dug into this answer, but it appears this is possible:

http://www.datatables.net/forums/discussion/1399/how-to-no-sort-by-default/p1

In DataTables 1.9 there was some aaSorting param that looks useful: http://legacy.datatables.net/ref

saimn commented:
Same as @taldcroft , it makes sense to disable the initial sort, and there is an option for this: `"order": []`
http://datatables.net/reference/option/order
Adding a column would allow to get back to the initial sort, it may be useful in some cases ?

eteq commented:
@taldcroft - Ah, good point that, regardless, it might be useful to disable the initial sort.

But I'm thinking this column might be useful anyway on the basis that if you sort them you want to know what the index is of that row (in fact, that's what prompted this issue in the first place: I sorted on a column and then wanted to grab the row *in python* to be able to do more with it).  But I don't want that in the "real" table because the only time I need it is when displaying in dataTables and its sorting functionality.

 I also agree with @saimn that there needs to be some way to "unsort" after you've sorted, and this fits the bill for that.  Although maybe a button that sets `order` to `[]` in the javascript also works for that?

taldcroft commented on astropy/table/table.py:
```diff
@@ -878,6 +878,10 @@ def show_in_notebook(self, tableid=None, css=None, display_length=50,
             to ``astropy.table.jsviewer.DEFAULT_CSS_NB``.
         display_length : int, optional
             Number or rows to show. Default to 50.
+        add_row_index : bool
```
Maybe using the word `show` instead of `add` here might make it more obvious that nothing is happening to the table itself, and instead this is just a feature of how the table is displayed.

taldcroft commented on astropy/table/table.py:
```diff
@@ -896,24 +900,30 @@ def show_in_notebook(self, tableid=None, css=None, display_length=50,
-                                max_lines=-1, show_dtype=False,
-                                tableclass=table_class)
+
+        has_row_index = 'row_index' in self.colnames
```
Although `row_index` is strictly more correct, I wonder if just `row` would suffice.  I'm bothered by the big chunk of wasted white space in the prime real estate on the left side of the table.

I also wonder if we could play silly games of adding space characters to the left and right of `row` until it is no longer in the original table `colnames`.

eteq replied:
I see your point re: `row_index`.  What do you think about `idx`?  That gets across the point that it's a python level index, not a "row" like in a spreadsheet (which is usually 1-indexed).

mhvk replied:
`seq`?

saimn replied:
Or `index` ?
I don't understand why, if `has_row_index` is True, you add the column but don't remove it. Is it possible to add a Column with the same name as an existing one ?

taldcroft replied:
Idx works for me.

eteq replied:
@saimn - ah, good catch, that's a bug - it definitely shouldn't be added when it's already there

taldcroft commented on astropy/table/table.py:
```diff
@@ -896,24 +900,30 @@ def show_in_notebook(self, tableid=None, css=None, display_length=50,
-                                tableclass=table_class)
+
+        has_row_index = 'row_index' in self.colnames
+        try:
```
I might suggest creating a temporary table.  This is cheap because of the `copy=False`, and maybe even faster in the end than adding a column and then removing it:
```
row_col_name = 'row'
while row_col_name in self.columns:
    row_col_name = ' ' + row_col_name

row_col = self.ColumnClass(name=row_col_name, data=np.arange(len(self)))
display_table = self.__class__([row_col] + self.columns.values(), copy=False)
html = display_table._base_repr(html=True, ...)
```

taldcroft commented on astropy/table/table.py:
```diff
@@ -958,9 +975,18 @@ def show_in_browser(self, max_lines=5000, jsviewer=False,
-                self.write(tmp, format='jsviewer', css=css,
-                           max_lines=max_lines, jskwargs=jskwargs,
-                           table_id=tableid, table_class=table_class)
+                has_row_index = 'row_index' in self.colnames
```
On frist glance it looks like this is a repeat of previous code and could be factored out.

taldcroft commented:
Overall I see the argument for having an explicit `row` column in the browser table, so I've just made some implementation suggestions assuming that is the case.

saimn commented:
Should this be enabled by default ? One use case I see is printing catalogs, which (almost) always have already an index column. Also it may be surprising for users to see an additional column without asking explicitly for it.

eteq commented:
Alright, @taldcroft @saimn, and @mhvk, the last set of commits should have addressed all your comments, as well as adding a test and changelog entry.  

Note that I opted for ``idx`` as that seemed a good compromise between "wanting to use up as little column space as possible" and "make it clear that this is the in-*python* index into the table rows"  That can be trivially changed if someone else feels strongly, though.

One thing to note: the test I added nominally requires IPython, so I added that to the conda dependencies.  I *could* modify things so that IPython is not required to do the tests, at the cost of some additional complexity/slightly less coverage.  But lets see how much longer this test build takes relative to *without* IPython (the previous one).  If it's only a small penalty, I prefer just including IPython.

eteq commented:
A quick check on the numbers before/after adding ipython suggests that it adds at most 30 sec to the build/install/test, which amounts to <~ 1% increase - that seems pretty trivial to me so I'm thinking this is good to go as-is.

And the tests should now be passing (the only failure was PEP8 freaking out on a minor thing that I just pushed up a fix to)

mhvk commented:
@eteq - this looks all OK to me, but I'll let @taldcroft have a last look.

taldcroft commented on astropy/table/table.py:
```diff
@@ -878,6 +886,10 @@ def show_in_notebook(self, tableid=None, css=None, display_length=50,
         display_length : int, optional
             Number or rows to show. Default to 50.
+        show_row_index : bool
+            If True, a column named "idx" will be added for display
```
Say that `True` is the default.

taldcroft commented on astropy/table/table.py:
```diff
@@ -856,9 +856,17 @@ def pprint(self, max_lines=None, max_width=None, show_name=True,
                 print(line)
 
+    def _make_index_row_display_table(self, index_row_name='idx'):
+        if index_row_name not in self.columns:
```
One thing you might do here is keep on adding a space to `index_row_name` until it isn't in the original table.  Since this is just a display column, I think the "hidden" spaces in the name wouldn't really be a problem.

In fact you could take this to the extreme and have the default `index_row_name` be `' '` (which is a perfectly legal column name).  Then it would end up looking like pandas notebook output where the row index column does not have a name label.

eteq replied:
Hmm, I see your point that you want to be able to get this sometimes, but that solution worries me for cases where "idx" is already in a table: I suspect the user (which is *me* in some cases I'm working with right now...) would get confused about which "idx" is the one that's actually in the table and which is auto-generated.  So I prefer that it gets hidden when "idx" is present.

I just pushed up what I think is probably a better solution anyway, though (see below).

taldcroft commented:
@eteq - I finished my review. thanks for working on this.  The ideas about the index column name are just ideas, take them if you like.   :smile:

eteq commented:
@taldcroft - after considering your suggestion here, I opted for a slightly different solution of instead making it *user-controllable* what the index column's name is.  Also, it clarifies that if the provided name already exists, `show_row_index` doesn't do anything (this is for compatibility in situations where for whatever reason you already have an index column) - I think that provides an optimal level of flexibility while still maintaining the convenience of just being able to do `tab.show_in_notebook()` without thinking too hard about it in most cases.

I'll go ahead and merge this when the tests pass then, as you indicated you're ok with it either way, @taldcroft - thanks!
